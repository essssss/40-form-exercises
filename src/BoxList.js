import { useState } from "react";
import Box from "./Box";
import NewBoxForm from "./NewBoxForm";
import { v4 as uuid } from "uuid";

const BoxList = () => {
    const initialBoxes = [];
    const [boxes, setBoxes] = useState(initialBoxes);

    const addBox = (newBox) => {
        setBoxes((boxes) => [...boxes, { ...newBox, id: uuid() }]);
    };
    const removeBox = (id) => {
        setBoxes((boxes) => boxes.filter((box) => box.id !== id));
    };

    return (
        <div>
            <NewBoxForm addBox={addBox} />
            {boxes.map(({ id, backgroundColor, width, height }) => (
                <Box
                    removeBox={removeBox}
                    id={id}
                    key={id}
                    backgroundColor={backgroundColor}
                    width={width}
                    height={height}
                />
            ))}
        </div>
    );
};

export default BoxList;
