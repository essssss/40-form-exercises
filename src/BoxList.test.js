import { render, fireEvent } from "@testing-library/react";
import BoxList from "./BoxList";

function addBox(boxList, height = "2", width = "2", color = "peachpuff") {
    const heightInput = boxList.getByLabelText("Height");
    const widthInput = boxList.getByLabelText("Width");
    const backgroundInput = boxList.getByLabelText("Box Color");
    fireEvent.change(backgroundInput, { target: { value: color } });
    fireEvent.change(widthInput, { target: { value: width } });
    fireEvent.change(heightInput, { target: { value: height } });
    const button = boxList.getByText("Submit");
    fireEvent.click(button);
}

it("should render", function () {
    render(<BoxList />);
});
it("matches snapshot", function () {
    const { asFragment } = render(<BoxList />);
    expect(asFragment()).toMatchSnapshot();
});

it("can add a new box", function () {
    const boxList = render(<BoxList />);

    // no boxes yet
    expect(boxList.queryByText("REMOVE BOX")).not.toBeInTheDocument();

    addBox(boxList);

    // expect to see a box
    const removeButton = boxList.getByText("REMOVE BOX");
    expect(removeButton).toBeInTheDocument();
    console.log(removeButton.previousSibling);
    expect(removeButton.previousSibling).toHaveStyle(`
	  width: 2em;
	  height: 2em;
	  background-color: peachpuff;
	`);
    // expect form to be empty
    expect(boxList.getAllByDisplayValue("")).toHaveLength(3);

    // expect(asFragment()).toMatchSnapshot();
});
