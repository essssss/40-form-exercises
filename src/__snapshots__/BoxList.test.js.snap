// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <div>
      <form>
        <label
          for="bg"
        >
          Box Color
        </label>
        <input
          id="bg"
          name="backgroundColor"
          placeholder="color"
          type="text"
          value=""
        />
        <label
          for="width"
        >
          Width
        </label>
        <input
          id="width"
          name="width"
          placeholder="width"
          type="text"
          value=""
        />
        <label
          for="height"
        >
          Height
        </label>
        <input
          id="height"
          name="height"
          placeholder="height"
          type="text"
          value=""
        />
        <button>
          Submit
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
